### PR TITLE
feat(treesitter): add `filetype` -> `lang` API

### PR DESCRIPTION
**Pull request by lewis6991**
## Problem

`vim.treesitter` does not know how to map a specific filetype to a parser.

This creates problems since in a few places ([including](https://github.com/neovim/neovim/pull/22191/commits/8a985d12dd6b4a5a4ba825939f36b7b1a324d849) in `vim.treesitter` itself), the `filetype` is incorrectly used in place of `lang`.

## Solution

Add an API to enable this:

- Add `vim.treesitter.language.add()` as a replacement for `vim.treesitter.language.require_language()`.
  - Optional arguments are now passed via an `opts` table.
  - Also takes a filetype (or list of filetypes) so we can keep track of what filetypes are associated with which langs.
  - Deprecated `vim.treesitter.language.require_language()`.
- Add `vim.treesitter.language.get_lang()` which returns the associated `lang` for a given `filetype`.
- Add `vim.treesitter.language.register()` to associate filetypes to a lang **without** loading the parser.

Note: This change works on the principle that `filetype` -> `lang` is a many to one relationship.

nvim-treesitter downstream PR: https://github.com/nvim-treesitter/nvim-treesitter/pull/4296

### CODE REVIEW

File: runtime/lua/vim/treesitter/language.lua [hunk 1]
```diff
@@ -23,15 +23,21 @@ function M.require_language(lang, path, silent, symbol_name)
       error("'" .. lang .. "' is not a valid language name")
     end
 
-    local fname = 'parser/' .. lang .. '.*'
-    local paths = a.nvim_get_runtime_file(fname, false)
-    if #paths == 0 then
+    for _, l in ipairs {lang, symbol_name} do
```
Comment (theHamsta) [inline runtime/lua/vim/treesitter/language.lua:26]: I'd rather have the user decide what they want instead of nvim trying out multiple paths. If the SO-filename should be configurable, it could be an extra parameter.

As I understood https://github.com/nvim-treesitter/nvim-treesitter/pull/4296#issuecomment-1426995523, this to enable `ft` to `lang` mapping like the Lua table that we have in nvim-treesitter. Please note that when you're moving the functionality of mapping `ft` to parser to the `treesitter.c` hashmap, you will load the same parser multiple times as the same SO-parser file can be loaded for multiple `lang`s (`lang` being the key for a loaded parser). Isn't this something that the caller of `require_language` can handle via a simple Lua table mapping? At nvim-treesitter, we always tried to have `ft` and `lang` be separate concepts.

Comment (lewis6991) [reply runtime/lua/vim/treesitter/language.lua:26]: > I'd rather have the user decide what they want instead of nvim trying out multiple paths. If the SO-filename should be configurable, it could be an extra parameter.

It already is as the second `path` argument.

> Please note that when you're moving the functionality of mapping ft to parser to the treesitter.c hashmap, you will load the same parser multiple times as the same SO-parser file can be loaded for multiple langs (lang being the key for a loaded parser).

I'm aware of this, and this is just a small implementation deficiency that should be easy enough to resolve. In practice, this will just mean that users with a `bash` and `zsh` buffer open in the same session will have the `sh` parser loaded multiple times. For the majority, this shouldn't be an issue since `filetype` usually maps directly to `lang`.

> At nvim-treesitter, we always tried to have ft and lang be separate concepts.

I've seen no benefit in keeping this distinction, and for me has only ever caused problems.

For example, I've currently had to introduce some awkward logic in https://github.com/neovim/neovim/pull/22191 because `get_parser` defaults to `filetype` for `lang` despite a parser already being associated for a buffer.

The end goal I'm working to is that the whole `filetype` -> `lang` relationship is encapsulated to parser installation and management. Everywhere else in any treesitter logic, only filetype should be considered.

In this PR I've only looked in both paths (for when `path` isn't provided) just to maintain some compatibility for users who will continue to pass `lang` as the first argument. Once downstream has adapted, which we won't give much time for, we then only search for parsers in `parser/{symbol_name}.so`

More context, I believe we were forced to maintain the `filetype` - `lang` distinction because of initial design mistakes in nvim's `treesitter.c` which was recently resolved with https://github.com/neovim/neovim/pull/19496.

**Comment by theHamsta** [comment]
I think the distinction between`lang` and `ft` is meaningful. `lang` is the decided by the parser developers outside of  the Neovim-context and `ft` is decided by Neovim. We had many weird bugs in the beginning of nvim-treesitter when parts of our code referred to `ft` and other to `lang`. This was especially useful when the file type was dynamically changed but could still use the same parser (js->react or bash ->PKGBUILD, long loading times when plugins changed 3x ft on file open). Multiple `ft` could have the same `lang` or you could switch the `lang` for one `ft`. This would map nicely to the internal C hashmap which holds one parser for each `<lang>`

I think the problem here could be the understanding what `lang` refers to. For me, `lang` refers to
- the grammar repo name `tree-sitter-<lang>`
- the grammar name in grammar.js and therefore the SO symbol
- `parser/<lang>.so` and `queries/<lang>/*.scm`

So basically the parser identification goes by the name the grammar author has chosen and have nvim-treesitter a configurable mapping from `ft` to `lang`. Of course you could also have 

- `parser/<ft>.so` and `queries/<ft>/*.scm`

and configure a mapping to the `.SO` symbol name which seems to be your understanding. If we want to to make `<lang>` and `<ft>` the same thing it would avoid confusion when we renamed `lang` in documentation and codebase. I think both approaches would work. It just has to be clear what `lang` means.

**Comment by lewis6991** [comment]
-  `parser/<ft>.so` and `queries/<ft>/*.scm`

I've not been clear about the direction yet, but this is not the intention. The `lang` is strictly tied to the parser and the symbol is actually completely configurable but 100% of the time is the same as `lang` currently. Most treesitter applications do not need to care or even have the information of `lang` and `get_parser()` is usually only used with a buffer number, and should always work. However right now it is broken because it defaults `lang` to the buffers filetype. To resolve this I believe we need `require_language()` to also be passed the filetypes the language is expected to handle.

Initially I planned to keep this filetype knowledge in the `treesitter.c` map, however on reflection (with help of your comment) I realised we can maintain this mapping in Lua and just tweak the signature of `require_language`.

I've updated the PR to add a `get_lang(filetype)` function which uses an internal map populated in `require_language`.

This means `get_parser` can default `lang` correctly and users of `require_language` can do:

```lua
require_language({'sh', 'csh', 'tcsh', 'bash', 'zsh'}, nil, nil, 'bash')
-- or
require_language('sh', nil, nil, 'bash')
require_language('csh', nil, nil, 'bash')
require_language('tcsh', nil, nil, 'bash')
require_language('bash', nil, nil, 'bash')
require_language('zsh', nil, nil, 'bash')
```

As opposed to just:

```lua
require_language('bash')
```

**Comment by clason** [comment]
Should the new function be used by `vim.treesitter.start()`? That is the point where an automatic filetype->lang lookup would be most useful.

Also, is the plan to deprecate the "old" functions?

**Comment by lewis6991** [comment]
> Should the new function be used by vim.treesitter.start()? That is the point where an automatic filetype->lang lookup would be most useful.

Sure.

> Also, is the plan to deprecate the "old" functions?

The plan is to at least deprecated `require_language`. Not sure about `inspect_language`.

**Comment by lewis6991** [comment]
I've added a `register` function so filetypes can be linked to a lang without loading a parser. This will means frameworks like `nvim-treesitter` can register the filetypes on `setup/config` without having to maintain their own API on top of core (which is aware of these mappings).

**Comment by lewis6991** [comment]
Marking ready, but I'm sure there are some aspect that might need iterating based on feedback.

File: runtime/lua/vim/treesitter/language.lua [hunk 2]
```diff
@@ -2,19 +2,66 @@ local a = vim.api
...
+local ft_to_lang = {}
+
+---@param filetype string
+---@return string|nil
+function M.get_lang(filetype)
+  return ft_to_lang[filetype]
+end
+
+---@deprecated
+function M.require_language(lang, path, silent, symbol_name)
```
Comment (justinmk) [inline runtime/lua/vim/treesitter/language.lua:15]: deprecate() ?

File: runtime/doc/treesitter.txt [hunk 3]
```diff
@@ -678,24 +708,12 @@ inspect_language({lang})          *vim.treesitter.language.inspect_language()*
...
         (table)
 
-                                  *vim.treesitter.language.require_language()*
-require_language({lang}, {path}, {silent}, {symbol_name})
-    Asserts that a parser for the language {lang} is installed.
-
-    Parsers are searched in the `parser` runtime directory, or the provided
-    {path}
+register({lang}, {filetype})              *vim.treesitter.language.register()*
+    Register a lang to be used for a filetype (or list of filetypes).
```
Comment (justinmk) [inline runtime/doc/treesitter.txt:712]: Is this something that would be done by filetype plugins? 

Any way to make this passive (via some file in 'runtimepath')? Is this a "many to many" mapping?

Comment (lewis6991) [reply runtime/doc/treesitter.txt:712]: > Is this something that would be done by filetype plugins?

Yes, and also nvim-treesitter / any parser management plugin. But also by the user, e.g: `vim.treesitter.language.register('bash', 'myniche_sh')`.

>  Is this a "many to many" mapping?

No, I want this to be a multiple filetypes to single lang mapping (mentioned in the OP). I really don't think there is a justifiable (or existing) use case for any filetype needing to use multiple parsers at the top level. Every exotic application I've seen would be handled by either: a complicated parser and/or language injections.

> Any way to make this passive (via some file in 'runtimepath')? Instead of "active" state managed at runtime.

The problem with the files in runtimepath method is that it makes it very difficult to redefine/override. I believe this is why `after/` was created. Also, I don't know what a "passive" would look like, though open to suggestions (or discussion). I can only think of a method where maybe `ftplugin` files have some directive to say what parser they should use (maybe setting a future `'lang'` option), and this would be set per buffer as opposed to what I've done here which keeps a separate side table.

File: runtime/doc/treesitter.txt [hunk 4]
```diff
@@ -678,24 +708,12 @@ inspect_language({lang})          *vim.treesitter.language.inspect_language()*
     Return: ~
         (table)
 
-                                  *vim.treesitter.language.require_language()*
-require_language({lang}, {path}, {silent}, {symbol_name})
-    Asserts that a parser for the language {lang} is installed.
-
-    Parsers are searched in the `parser` runtime directory, or the provided
-    {path}
+register({lang}, {filetype})              *vim.treesitter.language.register()*
```
Comment (justinmk) [inline runtime/doc/treesitter.txt:711]: Would an interface like `vim.treesitter.config()` similar to `vim.diagnostic.config()` make sense? Perhaps that helps discoverability.

File: runtime/doc/treesitter.txt (hunk 3 above)
Comment (theHamsta) [reply runtime/doc/treesitter.txt:712]: I think the many to one makes sense here. Many parsers for one filetype are still possible via injections. Our current model has a single root language but injections can even be added for the whole document

File: runtime/doc/treesitter.txt (hunk 4 above)
Comment (lewis6991) [reply runtime/doc/treesitter.txt:711]: Maybe. We could have something like: `vim.treesitter.config([lang}, {opts})`:

```lua
vim.treesitter.config('bash', {
  filetype = {'bash', 'sh'},
  ???
})
```

I think this only works though if a specific language can be configured once. This won't be ideal If a user also wants to further setup the language to also handle the `PKGBUILD` filetype.

However, I think most configuration points for treesitter are actually configured passively by the presence of query files: highlights, indent, folding.

As an example that this could be similar too, here's what I have to configure nvim-treesitter:

```lua
require'nvim-treesitter.configs'.setup {
  ensure_installed = {
    "bash", "c", "help", "html", "json", "lua", "make"
  },
  highlight = {
    enable = true,
  },
  indent = {
    enable = true,
    is_supported = function(lang)
      return ({
        lua = true,
        c = true,
        tcl = true
      })[lang] or false
    end
  },
  incremental_selection = {
    enable = true,
    keymaps = {
      init_selection    = "gnn",
      node_incremental  = "grn",
      scope_incremental = "grc",
      node_decremental  = "grm",
    },
  },
  fold = {
    enable = true,
    disable = {'rst', 'make'}
  },
  playground = { enable = true },
  context_commentstring = { enable = true },
  query_linter = {
    enable = true,
    use_virtual_text = true,
    lint_events = {"BufWrite", "CursorMoved"},
  },
}
```

However, I'm not sure this kind of config makes sense for a language provider.

Currently, a language provider provides:
- A parser (either directly or indirectly).
- A collection of queries under `queries/{lang}/`.
- Custom directives and predicates for queries (via `plugin/` I'd imagine).
- [This PR] a directive which says what filetypes the language aims to target.

Regardless, despite whether `vim.treesitter.config()` is a good idea, I think adding something like `register` should be fine as it is more low level (and can be removed if we change our minds).

Comment (theHamsta) [reply runtime/doc/treesitter.txt:711]: In nvim-treesitter, we also first had the approach that the user could provide a parser config with filetype associated. Internally, we had to invert the mapping similar way as it is done in this PR, because we actually we needed a filetype to lang mapping. Users sometimes had problems configuring the parser for one specific filetype. To do that they had to remove that filetype from the one parser config and add it to the other. Often they only added it to the alternative parser config and had then semi-random behavior (depending on which parser config registered first).

We ended up exposing the formally private map and discouraged setting filetypes on parser configs.

Comment (clason) [reply runtime/doc/treesitter.txt:711]: No, we very much don't want to copy nvim-treesitter's architecture! I think for core, it makes much more sense to be 
1. filetype-centric (i.e., you configure treesitter-based features in a ftplugin or autocommand, same as any other feature)
2. integrated into the `syntax` feature (which probably needs to be extended to cover the possibilities; e.g., by accepting a dict-like table -- once those are supported for options -- in addition to a boolean or string).

Comment (justinmk) [reply runtime/doc/treesitter.txt:711]: Can't config() start out with the impl of register() here? I'm hoping we can avoid the extra surface area. This is a configuration task, after all.

Comment (theHamsta) [reply runtime/doc/treesitter.txt:711]: > No, we very much don't want to copy nvim-treesitter's architecture! I think for core, it makes much more sense to be

I didn't say that. I just wanted to warn that `add(lang, {filetype = { "a", "b", "c" }})` can easily generate configuration errors when a users mentions ft `c"` for multiple languages. `register` does have this problem and would preferably be called in ftplugins so a filetype can select their parser rather than having parsers select their filetype. @lewis6991 already mentioned that `register` is preferred.

The implementation of nvim-treesitter of this is horrible. I'm happy when this PR can replace it.

Comment (lewis6991) [reply runtime/doc/treesitter.txt:711]: > Can't config() start out with the impl of register() here? I'm hoping we can avoid the extra surface area. This is a configuration task, after all.

The existing `config` model doesn't work well with multiple sources, so I'm unsure about introducing that right now without a more thought out design. I'd assume config would exist at `vim.treesitter.config()` (higher level) as opposed to `vim.treesitter.language.register()` (low level).

As far as surface area is concerned, I think a more pragmatic approach for treesitter should be more (but not totally) white-box for the time being, I.e. more things are accessible from the outside as this allows more experimentation from downstream. We can do this until we can figure out the exact shape we want things to take, and at that point we can begin to deprecate and private parts we want to unexpose. We may even need to iterate several times in some areas (like we already have). What's important is that we maintain progress.

My desired goal is that `register` becomes internal only and that `lang` becomes a mostly unimportant detail that's only needed when managing the parsers and queries.

>  register does have this problem and would preferably be called in ftplugins so a filetype can select their parser rather than having parsers select their filetype.

You're right. I was vaguely aware of this when putting this together and considered adding extra validation to detect such conflicts. For now, I decided on a passive approach of last wins, with the prospect of tightening this up in the future if it continues to be a problem. We have a similar problem with highlight groups (mostly solved with `default`), and syntax files (kind of solved with `after`).

Comment (justinmk) [reply runtime/doc/treesitter.txt:711]: > The existing `config` model doesn't work well with multiple sources, so I'm unsure about introducing that right now without a more thought out design. I'd assume config would exist at `vim.treesitter.config()` (higher level) as opposed to `vim.treesitter.language.register()` (low level).

ok agreed. 👍

File: runtime/lua/vim/treesitter/language.lua (hunk 2 above)
Comment (lewis6991) [reply runtime/lua/vim/treesitter/language.lua:15]: Our doc flow seems to only recognize `@deprecated`

Comment (lewis6991) [reply runtime/lua/vim/treesitter/language.lua:15]: Oh, you mean use `vim.deprecated()`

Comment (lewis6991) [reply runtime/lua/vim/treesitter/language.lua:15]: I added a call to `vim.deprecate` but I think the notify is a bit been obtrusive. We can add it later when at least nvim-treesitter adapts. `require_language` is just an alias for `add` anyway.

File: runtime/lua/vim/treesitter/health.lua [hunk 5]
```diff
@@ -17,7 +17,7 @@ function M.check()
 
   for _, parser in pairs(parsers) do
     local parsername = vim.fn.fnamemodify(parser, ':t:r')
-    local is_loadable, ret = pcall(ts.language.require_language, parsername)
```
Comment (clason) [inline runtime/lua/vim/treesitter/health.lua:20]: `require_language` returned `true` if no error occurred, unlike `add`. If this is intentional, the check in line 22 has to be adapted.
